### ISSUE TITLE
[Proposal] Avoid bound-checks on arrays, where applicable

### ISSUE THREAD

**Issue by Ultrahead** [dotnet/csharplang#530]
## Disable Bound-Check Operator

- [x] Proposed
- [ ] Prototype
- [ ] Implementation
- [ ] Specification

## Summary

Allow a way to tell the C# compiler that you want to avoid bound-checking when accessing a position on an array, whenever is possible.

## Motivation

By default, C# checks boundaries when accessing the values stored on arrays preventing off-by-one errors and buffer overflows. However, there are specific situations, when developers may need to disable said check to boost performance (for example, game developers) yet without threatening safety.

A common scenario could be something like ...:

```c#
...

for ( int i = 0; i < this.transitionTransforms.Length; i++ )
{
    fixed ( Matrix* currentMatrixPtr = &this.transitionTransforms[ i ] )
    {
        var targetMatrix = this.currentAnimation.BoneTransforms[ i ];
 
        currentMatrixPtr -> Decompose (out currentScale, out currentRotation, out currentTranslation);
        targetMatrix.Decompose (out targetScale, out targetRotation, out targetTranslation);

        Quaternion.Slerp (ref currentRotation, ref targetRotation, this.transitionProgress, out currentRotation);
        Vector3.Lerp (ref currentTranslation, ref targetTranslation, this.transitionProgress, out currentTranslation );
 
        *currentMatrixPtr = Matrix.CreateScale( currentScale ) * Matrix.CreateFromQuaternion( currentRotation ) * Matrix.CreateTranslation( currentTranslation );
    }
}

...
```
... which could also be refactored so as to use pointer arithmetic with a copy of currentMatrixPtr starting at position 0, instead of accessing the array on each iteration of the for block.

With the introdution of ref locals to c#, the above code could be rewritten as ...:

```c#
...

for ( int i = 0; i < this.transitionTransforms.Length; i++ )
{
    ref Matrix currentMatrix = ref this.transitionTransforms[ i ];
    var targetMatrix = this.currentAnimation.BoneTransforms[ i ];
 
    currentMatrix.Decompose (out currentScale, out currentRotation, out currentTranslation);
    targetMatrix.Decompose (out targetScale, out targetRotation, out targetTranslation);

    Quaternion.Slerp (ref currentRotation, ref targetRotation, this.transitionProgress, out currentRotation);
    Vector3.Lerp (ref currentTranslation, ref targetTranslation, this.transitionProgress, out currentTranslation );
 
    currentMatrix = Matrix.CreateScale( currentScale ) * Matrix.CreateFromQuaternion( currentRotation ) * Matrix.CreateTranslation( currentTranslation );
}

...
```

... which in turn brings the benefit of not having to mark the code as unsafe since we are not dealing with a pointer. But one perf issue remains, every time we retrieve a matrix from the array of transforms, the CLR checks for boundaries on the array even if the index variable "i" is valid unless the compiler catches the use case and deems it as safe even with no boundary-check. The problem with the latter, is that the developer does not know, not only whether the compiler is aware of the use case, but also whether it will generate bound-checkless IL code for it.

## Detail Design

Given the above-mentioned scenario, the proposal is to use the negation operator ("!") to let the compiler know that, for a specific case, you want it to not enforce bound checking when accessing a position on an array, if and only if it can assess and guarantee at compile time, that such access is deemed as secure.

```c#
...

for ( int i = 0; i < this.transitionTransforms.Length; i++ )
{
    ref Matrix currentMatrix = ref this.transitionTransforms![ i ]; // this request could be silently accepted by the compiler.
    var targetMatrix = this.currentAnimation.BoneTransforms![ i ]; // this request should be disregarded by the compiler either silently or with a warning.
 
    currentMatrix.Decompose (out currentScale, out currentRotation, out currentTranslation);
    targetMatrix.Decompose (out targetScale, out targetRotation, out targetTranslation);

    Quaternion.Slerp (ref currentRotation, ref targetRotation, this.transitionProgress, out currentRotation);
    Vector3.Lerp (ref currentTranslation, ref targetTranslation, this.transitionProgress, out currentTranslation );
 
    currentMatrix = Matrix.CreateScale( currentScale ) * Matrix.CreateFromQuaternion( currentRotation ) * Matrix.CreateTranslation( currentTranslation );
}

...
```
As it can be seen in the code above, the "!" operator is used to fetch the current matrix from transitionTransforms array. The idea here is that whenever the compiler finds such request, it must assess whether bound-checking can be skipped and properly proceed in consequence.

For the example, the compiler should skip bound-checking when accessing transtionTransforms![ i ], and enforce it for this.currentAnimation.BoneTransforms![ i ]. 

## Drawbacks

Compilation time will be increased every time a request is found in code, given that the compiler must determine, mandatorily, whether it can skip bound-checking for the use case.

## Alternatives 

One alternative could be to use the unsafe keyword to let the compiler know that you want to skip bound-checking, no matter what, even if you are not working with pointers. The drawback here, of course, is that the method must be marked as unsafe.

```c#
... unsafe method ...

for ( int i = 0; i < this.transitionTransforms.Length; i++ )
{
    ref Matrix currentMatrix = unsafe ref this.transitionTransforms![ i ]; // this request should be silently accepted by the compiler.
    var targetMatrix = unsafe this.currentAnimation.BoneTransforms![ i ]; // this request should be also silently accepted by the compiler since the method is unsafe.
 
    currentMatrix.Decompose (out currentScale, out currentRotation, out currentTranslation);
    targetMatrix.Decompose (out targetScale, out targetRotation, out targetTranslation);

    Quaternion.Slerp (ref currentRotation, ref targetRotation, this.transitionProgress, out currentRotation);
    Vector3.Lerp (ref currentTranslation, ref targetTranslation, this.transitionProgress, out currentTranslation );
 
    currentMatrix = Matrix.CreateScale( currentScale ) * Matrix.CreateFromQuaternion( currentRotation ) * Matrix.CreateTranslation( currentTranslation );
}

...
```
Please do notice the use of the keyword "unsafe" when accessing both, transtionTransforms![ i ] and this.currentAnimation.BoneTransforms![ i ]. 

## Unresolved Questions

Could this rationale also be extended to multidimensional and or jagged arrays?

## Design Meetings

n/a

**Comment by HaloFour** [comment]
So to TL;DR the above, it sounds like what you want is that if the C# compiler can ensure that the index will not be out of bounds that instead of emitting the `ld.elem.*` opcode to load an element from an array, which is where the CLR performs bounds checking that the compiler will instead emit `ldind` or `ldobj` opcodes to load indirectly from the raw address?

Doesn't the JIT already eliminate bounds checking in these simpler scenarios?  Does the use of these opcodes affect the "trustability" of the assembly?  For arbitrary structs is the cost of calculating the offset address less than bounds checking?

**Comment by Ultrahead** [comment]
If the Jitter does in fact eliminate bound-checking for "simpler" scenarions, then I don't think the C# would allow such optimization if it were to affect the "trustability" of the assembly.

Regarding arbitrary structs (and classes in general), it may, but at least in my case, I'd use this proposed feature with blittable structs with a defined layout and size.

**Comment by svick** [comment]
1. When you talk about "the compiler", do you mean the C# compiler or the JIT compiler? What IL do you expect to be emitted for `![]`?

2. I'm not sure I see the point of this feature. As far as I can tell, `![]` is telling the compiler to "try harder" when figuring out if it can avoid bound checks. But considering there is no definition of what "harder" means and that there is no guarantee that any bound checking will actually be avoided, this doesn't sound like something that would be worth having special syntax for.

   If it was some kind of annotation, similar to `[MethodImplAttribute(AggressiveInlining)]` (which is also not a guarantee of inlining), that would make more sense to me. But I'm not sure how would that apply here, at least not without something like Attributes everywhere (https://github.com/dotnet/roslyn/issues/6671).

3. In your example, it looks like `transitionTransforms` is a field. How could the bound checking be avoided there, considering the value of the field can change? Have you tried saving the array to a local variable and see if the existing JIT does avoid bound checking?

4. You're proposing the compiler should "try harder" (my words) for some specially marked indexing operations. Is there actually a room for improvement? And if there is, would such improvement actually be so slow it shouldn't be enabled either always, or based on some heuristic that would apply to all existing code?

**Comment by Ultrahead** [comment]
1. The C# compiler (and .NET Native). I expect the same IL emmited when no bound-check is used (so that both, the native compiler or the Jitter produce the same result).

2. It's more like "Try" (than "Try harder") and log a warning if disregarded (more in 4. below). I prefer ![ ] over an attribute because you can specify where to use it on a method.

3. "considering the value of the field can change". That depends on the use case. But the compiler can catch those situations and decide to enforce bound-checking. And then, when the warning is logged at compile time, I can decide a better solution, if any, and try again.

4. In any case, I prefer to decide my-self which approach to use per platform as a result of profiling. That is, either a) a local ref with [], b) a local ref with ![], a local ref with unsafe ![], or d) a pointer (for blittables only).

**Comment by HaloFour** [comment]
Also, `!` has also been suggested as a post-fix operator for insisting that a variable is not null when the compiler can't determine that.  So, assuming that said operator is implemented alongside non-nullable reference types, that syntax would already have a meaning.

**Comment by Ultrahead** [comment]
Yes, but I don't see it used like "![ ]" when accessing arrays.

But, I'm open to pick a different symbol to avoid the clash of meanings.

**Comment by HaloFour** [comment]
@Ultrahead 

The `!` operator and the `[]` operator could be combined in such a manner:

```cs
public int Foo(int[]? array) {
    // I know for a fact that array ain't null:
    int x = array![0];
    return x;
}
```

**Comment by Ultrahead** [comment]
Which symbol do you suggest, instead?

**Comment by HaloFour** [comment]
I don't have a suggested syntax.  In my opinion that's one of the last things that should need to be hammered out.  I'm more interested in discerning what such a feature actually does and whether it actually provides a benefit in enough use cases.  I'd have to see numbers.

**Comment by gafter** [comment]
It sounds like the proposed operator would have no effect on the meaning of the program. We tend not to add meaningless syntax to the language. Compilers (C# and jit jointly) already eliminate bounds checks when they can, and do not eliminate them when they cannot. Adding syntax to the source would be a placebo.

**Comment by Ultrahead** [comment]
Well, I bet many game devs that use C# (for example, with Unity) would love such a feature. I don't know its current install base, but I can estimate at a rough guess that it's way more than 1 million.

**Comment by Ultrahead** [comment]
#gafter: even with unsafe?

**Comment by fanoI** [comment]
I'd say that if AOT compiler / JIT does not eliminate bound checking in your example code it is a bug of the compiler... 'i' is computed is always on the array bound always between 0 and this.transitionTransforms.Length so the check must be omitted.

**Comment by Ultrahead** [comment]
What about `var targetMatrix = unsafe this.currentAnimation.BoneTransforms![ i ];`? Or any other case when you want to skip bound-checking no matter what?

**Comment by svick** [comment]
> What about […] any other case when you want to skip bound-checking no matter what?

Isn't that `unsafe` is for? How is the proposed `unsafe ![]` different from the existing `unsafe` functionality?

**Comment by Ultrahead** [comment]
No need for pointers. Available also to non-blittables.

**Comment by HaloFour** [comment]
@Ultrahead 

You'd still be using pointers and you'd still be held by the constraints of doing so, even if the syntax that you type doesn't contain a `*` character.

**Comment by Ultrahead** [comment]
No. You'd be using ref locals, which can be applied to non-blittables.

**Comment by HaloFour** [comment]
@Ultrahead 

Which, IIRC, you can't index into without `ldelem.*` and thus incurring the bounds checking.  If you want to load directly from a memory address offset, you'll need a pointer.

**Comment by Ultrahead** [comment]
Well, if this cannot be implemented for non-blittables in C#, then maybe it's time for me to move back to C++ for game dev projects, now that the GCC GNU compiler has added support for C++17.

**Comment by mikedn** [comment]
> If the Jitter does in fact eliminate bound-checking for "simpler" scenarios -as you imply, then I don't think the C# Team would allow such optimization in the first place if it were to affect the "trustability" of the assembly.

The JIT does eliminate range checks in the common case of
```C#
for (int i = 0; i < a.Length; i++) {
    // use a[i]
}
```
and a few other cases.

It's possible that it won't eliminate the range checks in your `transitionTransforms` case because it appears that `transitionTransforms` is a field. That's your problem, if you care about performance  then you should be a bit more careful with repeated field accesses.

In addition it's possible to use the Unsafe package to avoid some range checks even when non primitive types are involved. Granted, the resulting code would look a bit ugly but avoiding range checks shouldn't be a common case. If it is then perhaps you're using the wrong tool for the job.

**Comment by Ultrahead** [comment]
Wow. With that last phrase of yours, you have added a strong nail in the coffin, which I can picture it being used in slogans like:

"If you are using C#, then you are using the wrong tool for the job. ~Unreal Engine"

During many years I have suggested and or backed features meant to modernize C# and at the same time boost its performance to lower the distance with languages like C(99), C++ and D. To mention a few: primitives (blittables), generic pointers, enum constraints for generics, op constraints for generics (+,-, etc.), static and default implementations on interfaces, const ref parameters, const methods, and bound-checkless access for all types (including non-blittables).

Some of these have been around for years, and even though they are alive and kicking, they have been systematically pushed aside year by year for whatever reasons, being the most unusual one I have received here: "placebo", especially given the fact that some of the features added on latest versions of C# are syntactic sugar and not features that address performance improvements at all.

So, if the above-mentioned features, and even the possibility of supporting pointers for ref types that go along well with a GC (like D does, for example), do not get addressed soon, then I will sadly start to agree with you and, despite my love for C#, move on.

**Comment by MkazemAkhgary** [comment]
The branch predictor doesnt eliminate overhead of most likely to always true conditions? I think it does based on very famous answer in stack overflow. 

Special syntax would not help at all compared to what compilers and even further, branch predictors can do.

https://stackoverflow.com/questions/11227809/why-is-it-faster-to-process-a-sorted-array-than-an-unsorted-array

**Comment by MohammadHamdyGhanem** [comment]
I proposed to inhance foreach syntax so we can use it widly as a high level safe way to tell the compiler that no need for bound checks:
https://github.com/dotnet/csharplang/issues/1404

**Comment by vectorix** [comment]
Why don't you use Unsafe.Add?

**Comment by Trayani** [comment]
Since the issue is still open, I take the liberty of adding my thoughts :-).
The original proposal kind of aimed to let the compiler do a deeper evaluation of the necessity of bounds check which was dismissed. 
It was suggested to use of pointer or Unsafe.Add. What I dislike about it is that we have to pin objects that are accessed from the array using `fixed` which can be counter-productive performance-wise. It just seems like a band-aid.

Wouldn't be the easiest solution to let us have a method Attribute, e.g. [NoArrayBoundsCheck], where we basically declare
"I'm 100 % sure all indexes will be within bounds and I accept the responsibility for potential errors. Now, there, there compiler, save that bounds check for another time". ?
The attribute could be "hidden" in System.Runtime.CompilerServices from inexperienced devs, produce warnings or even require `unsafe` wrapper.
 @svick already mentioned this above. Though in my opinion, the potential attribute does not need to apply to a more specific location than a method (but it would be nice to have, sure).

**Comment by svick** [comment]
@Trayani 

> It was suggested to use a pointer or Unsafe.Add. What I dislike about it, is that we have to pin objects that are accessed from the array using `fixed` which can be counter-productive performance-wise. Honestly, It just seems like a band-aid to me (the issue is not being directly addressed).

There's also an overload of `Unsafe.Add` that works with `ref`s:

```c#
public static ref T Add<T> (ref T source, int elementOffset)
```

From a quick look at [disassembly of code that uses it](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAEzkuygN7Ly+8qxYBtALrlsipQfJ6DRjOVHs+wgBQBKO/t3Wb9lnJcBnxyAF5yUit/AwAzaHI3B0NwyLkUgB4LGgAZGCYAcwwAC3TWAGpyrz8Ym2DQ8ojsY1ZTaNriAHYgkPb9AF9vWxryZlxsWJhDQIAFCAcRTyHfWunHetSooZtY1gQYABNEhwAqcj5U7GrVnYSkwNZNsvIs7Fz8otLDSuubmI3GucWm1tgYuj18H1yIMVCNkgBVJjjSYAQQOBzoMFiSxGK1qyQ2ES2IxssFia3IUFSZIsxlIIJJcTuyUeROer3ehRKZR+oP+IXIgMRyM4aIObhpUDQhg8UJs4PqUJh/SAA==), it looks like the code is as efficient as the `unsafe` version, but does not require pinning.

**Comment by Trayani** [comment]
@svick 
You got me there. I was not aware that `fixed `does not have to be used with a class-typed `ref ` parameter. Well, that counters my prior arguments =).

**Comment by reflectronic** [comment]
The `no.` IL prefix (§III.2.2), which allows null checks/bounds checks/type checks to be skipped for certain opcodes, could be used to implement such a feature. It isn't currently implemented in CoreCLR though, so that would need to be done first.

**Comment by Ultrahead** [comment]
Closing ...

**Comment by abelbraaksma** [comment]
@reflectronic, that may change, see: https://github.com/dotnet/runtime/issues/10112 (which, if implemented, would provide a way to do what this issue was about in the first place, though for the time being, would require emitting IL yourself, or use Cecil).

**Comment by Ultrahead** [comment]
Although I closed this issue I would like to state here that I love what's coming in C# 9 in terms of performance and interop: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9

Like function pointers and native-size integers.

For those of us who dare use C# for game development and even game-engine development, this is good news.

**Comment by Modus-Exciter** [comment]
Such feature is needed, when the array is used inside another class with more strict bounds requirements, than array has by itself. For example, I use triangle bit matrixes based on the BitArray class. In my Matrix class I must check both indexes, because when one of them is bigger than matrix size, the second one can be small, and the linear index based on them can be inside the BitArray bounds. The second check is performed by the BitArray class, because its length is not necessary a multiple of 32. The third check is performed by the int array in side the BitArray object. I'm able to avoid only the second of these checks implementing BitArray by my own. But I really need only one check on the highest level, because it's the most strict. There is the similar situation in all standard collections based on arrays, such as List, Dictionary, HashSet, SortedList, Queue and Stack. They contain an array which is used definently correctly. Whithout array bounds check they would work faster. I've made performance checks, it can increase performance from 20 to 150 per cent. Isn't it enough?

**Comment by dynamite7** [comment]
> From a quick look at [disassembly of code that uses it](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAEzkuygN7Ly+8qxYBtALrlsipQfJ6DRjOVHs+wgBQBKO/t3Wb9lnJcBnxyAF5yUit/AwAzaHI3B0NwyLkUgB4LGgAZGCYAcwwAC3TWAGpyrz8Ym2DQ8ojsY1ZTaNriAHYgkPb9AF9vWxryZlxsWJhDQIAFCAcRTyHfWunHetSooZtY1gQYABNEhwAqcj5U7GrVnYSkwNZNsvIs7Fz8otLDSuubmI3GucWm1tgYuj18H1yIMVCNkgBVJjjSYAQQOBzoMFiSxGK1qyQ2ES2IxssFia3IUFSZIsxlIIJJcTuyUeROer3ehRKZR+oP+IXIgMRyM4aIObhpUDQhg8UJs4PqUJh/SAA==), it looks like the code is as efficient as the `unsafe` version, but does not require pinning.

Instead of even checking the first element is possible to completely remove the checks with:
Unsafe.Add(ref System.Runtime.InteropServices.MemoryMarshal.GetArrayDataReference(a), i);

**Comment by GabrielMotaAlexandre** [comment]
> From a quick look at [disassembly of code that uses it](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAEzkuygN7Ly+8qxYBtALrlsipQfJ6DRjOVHs+wgBQBKO/t3Wb9lnJcBnxyAF5yUit/AwAzaHI3B0NwyLkUgB4LGgAZGCYAcwwAC3TWAGpyrz8Ym2DQ8ojsY1ZTaNriAHYgkPb9AF9vWxryZlxsWJhDQIAFCAcRTyHfWunHetSooZtY1gQYABNEhwAqcj5U7GrVnYSkwNZNsvIs7Fz8otLDSuubmI3GucWm1tgYuj18H1yIMVCNkgBVJjjSYAQQOBzoMFiSxGK1qyQ2ES2IxssFia3IUFSZIsxlIIJJcTuyUeROer3ehRKZR+oP+IXIgMRyM4aIObhpUDQhg8UJs4PqUJh/SAA==), it looks like the code is as efficient as the `unsafe` version, but does not require pinning.

Instead of even checking the first element it's possible to completely remove the checks with:

```cs
Unsafe.Add(ref System.Runtime.InteropServices.MemoryMarshal.GetArrayDataReference(a), i);
```

**Comment by radfast** [comment]
I'd like to eliminate bounds checking under the programmer's control using `Assert()` statements, something like this for example:

    int i, count;
    int sum = 0;

    // some code which calculates i and count in some way the runtime does not follow when optimising out bounds checks

    Assert(i > 0);
    Assert(count <= array.Length - 1);
    for ( ; i < count; i+=2)
    {
        sum += array[i] * array[i + 1];
    }

I appreciate that `Assert()` is a debug-mode only command, ignored in Release builds.  But something like this would be entirely satisfactory and would solve the issue, in practice, for most coders

**Comment by CyrusNajmabadi** [comment]
@radfast this is a closed issue.  if you want a new issue on this, open a discussion on dotnet/csharplang.  note though, you can trivially avoid bounds checks.  Just fix the array into a pointer, and access the pointer.  It's unsafe, but that's exactly what you're asking for.
